File: app/src/main/java/souch/smp/MergeAudioProcessor.java
package souch.smp;

import android.util.Log;

import androidx.media3.common.util.UnstableApi;

import java.nio.ByteBuffer;

@UnstableApi
public class MergeAudioProcessor extends BypassAudioProcessor {
    private boolean stereo = true;

    @Override
    public boolean handleBuffer(ByteBuffer inputBuffer) {
        if (!support || stereo) return false;

        while (inputBuffer.remaining() >= 4) {
            int l = inputBuffer.getShort();
            int r = inputBuffer.getShort();
            short avg = (short) ((l + r) / 2);
            outputBuffer.putShort(avg);
            outputBuffer.putShort(avg);
        }
        outputBuffer.flip();
        return true;
    }

    public boolean isStereo() {
        return stereo;
    }

    public void setStereo(boolean stereo) {
        this.stereo = stereo;
    }
}

File: app/src/main/java/xyz/mordorx/sicmu/media/MergeAudioProcessor.kt
package xyz.mordorx.sicmu.media

import androidx.media3.common.util.UnstableApi
import java.nio.ByteBuffer

/** This can downmix stereo into mono in real-time.
 * 
 * @author LoliBall ([...](https://github.com/WhichWho)) on 2022-12-22
 */
@UnstableApi
class MergeAudioProcessor : BypassAudioProcessor() {
    var isStereo: Boolean = true

    public override fun handleBuffer(inputBuffer: ByteBuffer): Boolean {
        if (!support || this.isStereo) return false

        while (inputBuffer.remaining() >= 4) {
            val l = inputBuffer.getShort().toInt()
            val r = inputBuffer.getShort().toInt()
            val avg = ((l + r) / 2).toShort()
            outputBuffer.putShort(avg)
            outputBuffer.putShort(avg)
        }
        outputBuffer.flip()
        return true
    }
}
